// keystone/admin/client/components/Navigation/PrimaryNavigation.js
import React from 'react';
import { Container } from 'elemental';

var PrimaryNavItem = React.createClass({
	displayName: 'PrimaryNavItem',
	propTypes: {
		children: React.PropTypes.node.isRequired,
		className: React.PropTypes.string,
		href: React.PropTypes.string.isRequired,
		label: React.PropTypes.string,
		title: React.PropTypes.string,
	},
	render () {
		return (
			<li className={this.props.className} data-section-label={this.props.label}>
				<a href={this.props.href} title={this.props.title} tabIndex="-1">
					{this.props.children}
				</a>
			</li>
		);
	},
});

var PrimaryNavigation = React.createClass({
	displayName: 'PrimaryNavigation',
	propTypes: {
		brand: React.PropTypes.string,
		currentSectionKey: React.PropTypes.string,
		sections: React.PropTypes.array.isRequired,
		signoutUrl: React.PropTypes.string,
	},
	getInitialState () {
		return {};
	},
	componentDidMount () {
		this.handleResize();
		window.addEventListener('resize', this.handleResize);
	},
	componentWillUnmount () {
		window.removeEventListener('resize', this.handleResize);
	},
	handleResize () {
		this.setState({
			navIsVisible: window.innerWidth >= 768,
		});
	},
	renderSignout () {
		if (!this.props.signoutUrl) return null;

		return (
			<PrimaryNavItem label="octicon-sign-out" href={this.props.signoutUrl} title="Sign Out">
				<span className="octicon octicon-sign-out" />
			</PrimaryNavItem>
		);
	},
	renderFrontLink () {
		return (
			<ul className="app-nav app-nav--primary app-nav--right">
				<PrimaryNavItem label="octicon-globe" href={Keystone.backUrl} title={'Front page - ' + this.props.brand}>
					<span className="octicon octicon-globe" />
				</PrimaryNavItem>
				{this.renderSignout()}
			</ul>
		);
	},
	renderBrand () {
		// TODO: support navbarLogo from keystone config
		return (
			<PrimaryNavItem label="octicon-home" className={this.props.currentSectionKey === 'dashboard' ? 'active' : null} href={Keystone.adminPath} title={'Dashboard - ' + this.props.brand}>
				<span className="octicon octicon-home" />
			</PrimaryNavItem>
		);
	},
	renderNavigation () {
		if (!this.props.sections || !this.props.sections.length) return null;

		return this.props.sections.map((section) => {
			const href = section.lists[0].external ? section.lists[0].path : `${Keystone.adminPath}/${section.lists[0].path}`;
			const className = (this.props.currentSectionKey && this.props.currentSectionKey === section.key) ? 'active' : null;

			return (
				<PrimaryNavItem key={section.key} label={section.label} className={className} href={href}>
					{section.label}
				</PrimaryNavItem>
			);
		});
	},
	render () {
		if (!this.state.navIsVisible) return null;

		return (
			<nav className="primary-navbar">
				<Container clearfix>
					<ul className="app-nav app-nav--primary app-nav--left">
						<li data-section-label="octicon-home" data-reactid=".0.0.1.0.1:1.0"><a href="/keystone" title="Dashboard - Congo Red Dot" tabindex="-1" data-reactid=".0.0.1.0.1:1.0.0"><span class="octicon octicon-home" data-reactid=".0.0.1.0.1:1.0.0.0"></span></a></li>
						<li data-section-label="Upload" data-reactid=".0.0.1.0.1:0.2:$upload"><a href="/keystone/galleries" tabindex="-1" data-reactid=".0.0.1.0.1:0.2:$upload.0">Upload</a></li>
						<li data-section-label="Test" data-reactid=".0.0.1.0.1:0.2:$test"><a href="/test" tabindex="-1" data-reactid=".0.0.1.0.1:0.2:$test.0">Test</a></li>
						<li data-section-label="Results" data-reactid=".0.0.1.0.1:0.2:$results"><a href="/keystone/results" tabindex="-1" data-reactid=".0.0.1.0.1:0.2:$results.0">Results</a></li>
						<li data-section-label="Howto" data-reactid=".0.0.1.0.1:0.2:$howto"><a href="/howto" tabindex="-1" data-reactid=".0.0.1.0.1:0.2:$howto.0">Howto</a></li>
					</ul>
					{this.renderFrontLink()}
				</Container>
			</nav>
		);
	},
});

module.exports = PrimaryNavigation;
